proxy/ParentYamlConfig.cc

```diff
@@ -0,0 +1,224 @@
+    p.dest_domain = node["dest_domain"].as<std::string>();
+  }
+  if (node["dest_host"]) {
+    p.dest_host = node["dest_host"].as<std::string>();
+  }
+  if (node["dest_ip"]) {
+    p.dest_ip = node["dest_ip"].as<std::string>();
+  }
+  if (node["url_regex"]) {
+    p.url_regex = node["url_regex"].as<std::string>();
+  }
+  if (node["scheme"]) {
+    p.scheme = node["scheme"].as<std::string>();
+  }
+  if (node["prefix"]) {
+    p.prefix = node["prefix"].as<std::string>();
+  }
+  if (node["suffix"]) {
+    p.suffix = node["suffix"].as<std::string>();
+  }
+  if (node["method"]) {
+    p.method = node["method"].as<std::string>();
+  }
+  if (node["time"]) {
+    p.time = node["time"].as<std::string>();
+  }
+  if (node["src_ip"]) {
+    p.src_ip = node["src_ip"].as<std::string>();
+  }
+  if (node["parent"]) {
+    p.parent = node["parent"].as<std::string>();
+  }
+  if (node["secondary_parent"]) {
+    p.secondary_parent = node["secondary_parent"].as<std::string>();
+  }
+  if (node["parent_retry"]) {
+    p.parent_retry = node["parent_retry"].as<std::string>();
+  }
+  if (node["unavailable_server_retry_responses"]) {
+    p.unavailable_server_retry_responses = node["unavailable_server_retry_responses"].as<std::string>();
+  }
+  if (node["round_robin"]) {
+    p.round_robin = node["round_robin"].as<std::string>();
+  }
+  if (node["qstring"]) {
+    p.qstring = node["qstring"].as<std::string>();
+  }
+
+  // get all the int's
+  if (node["port"]) {
+    p.port = node["port"].as<int>();
+  }
+  if (node["secondary_mode"]) {
+    p.secondary_mode = node["secondary_mode"].as<int>();
+  }
+  if (node["max_simple_retries"]) {
+    p.max_simple_retries = node["max_simple_retries"].as<int>();
+  }
+  if (node["max_unavailable_server_retries"]) {
+    p.max_unavailable_server_retries = node["max_unavailable_server_retries"].as<int>();
+  }
+
+  // get all the booleans
+  if (node["internal"]) {
+    p.internal = node["internal"].as<bool>();
+  }
+  if (node["parent_is_proxy"]) {
+    p.parent_is_proxy = node["parent_is_proxy"].as<bool>();
+  }
+  if (node["go_direct"]) {
+    p.go_direct = node["go_direct"].as<bool>();
+  }
+
+  return p;
+}
+
+int
+ParentYamlConfig::Init(const std::string &content)
+{
+  YAML::Node root;
+  YAML::Node destinations;
+
+  try {
+    root = YAML::Load(content);
+
+    if (!root.IsMap()) {
+      error_msg << __FILE__ << ":" << __LINE__ << ": "
+                << "yaml parse error, expected a map at the root node.";
+      return 1;
+    }
+    destinations = root["destinations"];
+    if (!destinations.IsSequence()) {
+      error_msg << __FILE__ << ":" << __LINE__ << ": "
+                << "yaml parse error, expected a 'destinations' node of type sequence.";
+      return 1;
+    } else {
+      dest_size = destinations.size();
+    }
+    for (uint i = 0; i < dest_size; i++) {
+      parent_dest *p = new parent_dest();
+      destinations[i] >> *p;
+      if (verifyConfig(*p, i + 1)) {
```

Comment:
This seems to leak a dynamic instance of `parent_dest` if `verifyConfig()` returns false.  It would also leaks if the >> operator or `verifyConfg()` throw an exception.

Reply:
Look at `std::unique_ptr` - you can `release` at the end when everything is OK.

Reply:
If you use release instead of std::move when adding the unique_ptr to the container, you'd want to use push_back not emplace_back.  This will create a temporary unique_ptr, and call the overload of push_back whose parameter is an rvalue reference.  This would ensure that the object gets deleted if growing the container causes an out-of-heap-memory exception to be thrown.  Although, in ATS, I think we just directly terminate if we run out of heap.

Reply:
no-longer using this in the PR, outdated.